File: src/main/java/com/azbj/proposalde3/service/QuestionnaireService.java
package com.azbj.proposalde3.service;

import com.azbj.proposalde3.repository.QuestionnaireRepository;
import com.azbj.proposalde3.model.Questionnaire;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class QuestionnaireService {

    @Autowired
    private QuestionnaireRepository questionnaireRepository;

    private String v_isalch;

    public List<Questionnaire> getQuestionnaireDetails() {
        return questionnaireRepository.findAll();
    }

    public void updateVIsalch(String ANS1) {
        if ("Y".equals(ANS1)) {
            v_isalch = "Y";
        } else if ("N".equals(ANS1)) {
            v_isalch = "N";
        } else {
            v_isalch = "I";
        }
    }

    public boolean checkFreezeQuestFlag() {
        // Logic to check if the 'FREEZE_QUEST' flag is enabled
        return false; // Placeholder return value
    }

    public void callProcedureBasedOnGrpProduct(String v_grp_product, String gender, String part_id, String name) {
        // Logic to call appropriate procedure based on 'v_grp_product'
    }

    public int logOnlineActivities(String application_no) {
        // Logic to count the number of online activities related to the application number
        return 0; // Placeholder return value
    }

    public void handleException(Exception e) {
        // Logic to log the error message
        e.printStackTrace();
    }

    public String validateResponse(int questionId, int subQuestionId, String response) {
        // Logic to validate the response based on the given criteria
        return ""; // Placeholder return value
    }

    public void processResponse(int questionId, int subQuestionId, String response) {
        // Logic to process the response and update the status variables
    }

    public String getProposalTypeAndPages(String applicationNumber) {
        // Logic to determine the proposal type and number of pages based on the application number
        return ""; // Placeholder return value
    }

    public List<String> fetchImages(int questionId, int subQuestionId) {
        // Logic to fetch specific images based on the question ID and sub-question values
        return null; // Placeholder return value
    }

    public String validateCorrectValue(int questionId, int subQuestion, String correctValue) {
        if (questionId == 6 && subQuestion == 5 && correctValue == null) {
            return "Husband Insurance cannot be null";
        } else if (questionId == 6 && subQuestion == 0) {
            if ("Y".equals(correctValue)) {
                return "Y";
            } else if ("N".equals(correctValue)) {
                return "N";
            } else {
                return "I";
            }
        } else if (questionId == 6 && subQuestion == 1) {
            if ("I".equals(correctValue)) {
                return "Please check pregnancy status.";
            } else if ("N".equals(correctValue) && correctValue != null) {
                return "She is not pregnant... Invalid Entry.";
            } else if ("Y".equals(correctValue) && (Integer.parseInt(correctValue) < 1 || Integer.parseInt(correctValue) > 12)) {
                return "The Pregnancy months can be between 1 and 12.";
            } else if ("Y".equals(correctValue) && correctValue == null) {
                return "Please enter the month of pregnancy.";
            }
        } else if (questionId == 4 && subQuestion == 2) {
            if ("Y".equals(correctValue)) {
                return "Y";
            } else if ("N".equals(correctValue)) {
                return "N";
            } else {
                return "I";
            }
        } else if (questionId == 4 && (subQuestion == 3 || subQuestion == 4 || subQuestion == 5)) {
            if ("I".equals(correctValue)) {
                return "Please check Alcoholism Question.";
            } else if ("N".equals(correctValue) && correctValue != null) {
                return "The person does not drink alcohol... Invalid answer.";
            } else if ("Y".equals(correctValue) && correctValue == null) {
                return "Please give in details of the drinking habits.";
            }
        }
        return ""; // Placeholder return value
    }

    public void validateResponses(int questionId, int subQuestionId, String correctValue) {
        // Logic to validate the responses based on the specified business rules
    }

    public void populateQuestions() {
        // Logic to fetch questions based on criteria, delete specific records based on conditions, and handle errors
    }

    public void checkSystemConstants(String newPaymentMode) {
        // Logic to perform a database query to check if there are any system constants with type 'SCB' and code 'SCB_PAYMODE' that match the FSC code and are not expired
    }

    public void checkApplicationBypassDetails(String applicationNo) {
        // Logic to perform a database query to check if there are any application bypass details for the given application number with a bypass flag set to 'Y'
    }

    public void checkCardCases(String applicationNo) {
        // Logic to perform a database query to check if there are any card cases for the given application number with a bank reference code containing 'SCBCC'
    }

    public void processITDeclaration(String applicationNumber, String insuredPersonNumber, String contractID, String mandatoryStatus, String passportProof, String nationality, String ipNationality, String propertyType) {
        // Logic to process the IT Declaration data
    }

    public void deleteExistingRecords(String applicationNumber, String proposalStatus, int questionId, int subQuestion) {
        // Logic to delete existing records in the `azbj_fcf_questionnaire` table
    }

    public void insertNewRecords(List<Questionnaire> questionnaireRecords) {
        // Logic to insert new records into the `azbj_fcf_questionnaire` table
    }

    public void handleRenewalPaymentModeChange(String newPaymentMode) {
        // Logic to handle renewal payment mode changes and perform database checks
    }

    public void handleMonthlyRenewalFrequency(String frequency) {
        // Logic to populate questions based on the renewal frequency
    }
}

File: src/app/services/questionnaire.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Questionnaire } from '../models/questionnaire.model';

@Injectable({
  providedIn: 'root'
})
export class QuestionnaireService {
  private apiUrl = 'http://localhost:8080/api/questionnaire';

  constructor(private http: HttpClient) {}

  getQuestionnaireDetails(): Observable<Questionnaire[]> {
    return this.http.get<Questionnaire[]>(`${this.apiUrl}/details`);
  }

  updateStatus(questionId: number, subQuestion: number, correctValue: string): string {
    if (questionId === 6 && subQuestion === 0) {
      if (correctValue === 'Y') return 'Y';
      if (correctValue === 'N') return 'N';
      return 'I';
    }
    return '';
  }

  updateVIsalch(ANS1: string): string {
    if (ANS1 === 'Y') return 'Y';
    if (ANS1 === 'N') return 'N';
    return 'I';
  }

  checkFreezeQuestFlag(): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/checkFreezeQuestFlag`);
  }

  callProcedureBasedOnGrpProduct(v_grp_product: string, gender: string, part_id: string, name: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/callProcedureBasedOnGrpProduct`, { v_grp_product, gender, part_id, name });
  }

  logOnlineActivities(application_no: string): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/logOnlineActivities/${application_no}`);
  }

  validateResponse(questionId: number, subQuestionId: number, response: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/validateResponse`, { questionId, subQuestionId, response });
  }

  loadQuestionnaireImage(imageId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/loadQuestionnaireImage/${imageId}`);
  }

  processResponse(questionId: number, subQuestionId: number, response: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/processResponse`, { questionId, subQuestionId, response });
  }

  getProposalTypeAndPages(applicationNumber: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/getProposalTypeAndPages/${applicationNumber}`);
  }

  fetchImages(questionId: number, subQuestionId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/fetchImages/${questionId}/${subQuestionId}`);
  }

  validateCorrectValue(questionId: number, subQuestion: number, correctValue: string): string {
    if (questionId === 6 && subQuestion === 5 && !correctValue) return 'Husband Insurance cannot be null';
    if (questionId === 6 && subQuestion === 0) {
      if (correctValue === 'Y') return 'Y';
      if (correctValue === 'N') return 'N';
      return 'I';
    }
    if (questionId === 6 && subQuestion === 1) {
      if (correctValue === 'I') return 'Please check pregnancy status.';
      if (correctValue === 'N' && correctValue) return 'She is not pregnant... Invalid Entry.';
      if (correctValue === 'Y' && (correctValue < '1' || correctValue > '12')) return 'The Pregnancy months can be between 1 and 12.';
      if (correctValue === 'Y' && !correctValue) return 'Please enter the month of pregnancy.';
    }
    if (questionId === 4 && subQuestion === 2) {
      if (correctValue === 'Y') return 'Y';
      if (correctValue === 'N') return 'N';
      return 'I';
    }
    if (questionId === 4 && [3, 4, 5].includes(subQuestion)) {
      if (correctValue === 'I') return 'Please check Alcoholism Question.';
      if (correctValue === 'N' && correctValue) return 'The person does not drink alcohol... Invalid answer.';
      if (correctValue === 'Y' && !correctValue) return 'Please give in details of the drinking habits.';
    }
    return '';
  }

  validateResponses(responses: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/validateResponses`, responses);
  }

  validateAnswers(ans1: string, ans_de1: string, correct: number): string {
    if (![1, 2, 3, 4, 7, 8].includes(correct)) return 'Invalid value entered';
    if (correct === 3 && ans1 !== ans_de1) return 'DE1 and DE2 Values Are Not Matching';
    if ([1, 2].includes(correct) && ans1 === ans_de1) return 'Please select 3, 4, 7 or 8';
    if ([3, 2].includes(correct)) return ans1;
    if (correct === 1) return ans_de1;
    return '';
  }

  validateQuestionnaireResponse(questionId: number, subQuestionId: number, correctValue: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/validateQuestionnaireResponse`, { questionId, subQuestionId, correctValue });
  }

  getQuestions(): Observable<Questionnaire[]> {
    return this.http.get<Questionnaire[]>(`${this.apiUrl}/getQuestions`);
  }

  getQuestionsByCriteria(partnerId: string, applicationNumber: string): Observable<Questionnaire[]> {
    return this.http.post<Questionnaire[]>(`${this.apiUrl}/getQuestionsByCriteria`, { partnerId, applicationNumber });
  }

  deleteQuestion(questionId: string, subQuestion: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/deleteQuestion/${questionId}/${subQuestion}`);
  }

  getExcludedQuestions(productCode: string): Observable<Questionnaire[]> {
    return this.http.get<Questionnaire[]>(`${this.apiUrl}/getExcludedQuestions/${productCode}`);
  }

  populateQuestions(): Observable<Questionnaire[]> {
    return this.http.get<Questionnaire[]>(`${this.apiUrl}/populateQuestions`);
  }

  updateAnswer(questionId: number, answerType: string, correctValue: string, answerDescription: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/updateAnswer`, { questionId, answerType, correctValue, answerDescription });
  }

  navigateToQualityCheck(): void {
    // Logic to navigate to quality check questions section
  }

  setFocusToAnswerField(): void {
    // Logic to set focus to the answer field in quality check questions
  }

  fetchDependentQuestions(questionId: string, answer: string): Observable<Questionnaire[]> {
    return this.http.post<Questionnaire[]>(`${this.apiUrl}/fetchDependentQuestions`, { questionId, answer });
  }

  deleteDependentQuestions(questionId: string, answer: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/deleteDependentQuestions`, { questionId, answer });
  }

  submitITDeclaration(applicationNumber: string, insuredPersonNumber: string, contractID: string, mandatoryStatus: string, passportProof: string, nationality: string, ipNationality: string, propertyType: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/submitITDeclaration`, { applicationNumber, insuredPersonNumber, contractID, mandatoryStatus, passportProof, nationality, ipNationality, propertyType });
  }

  deleteExistingRecords(applicationNumber: string, proposalStatus: string, questionId: number, subQuestion: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/deleteExistingRecords`, { applicationNumber, proposalStatus, questionId, subQuestion });
  }

  insertNewRecords(records: any[]): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/insertNewRecords`, records);
  }

  checkSystemConstants(newPaymentMode: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/checkSystemConstants`, { newPaymentMode });
  }

  checkApplicationBypassDetails(applicationNo: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/checkApplicationBypassDetails`, { applicationNo });
  }

  checkCardCases(applicationNo: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/checkCardCases`, { applicationNo });
  }

  handleMonthlyRenewalFrequency(frequency: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/handleMonthlyRenewalFrequency`, { frequency });
  }
}
